### ISSUE TITLE
How to make an inference using pretrained models?

### ISSUE THREAD

**Issue by dimasheva1** [meilfang/LMFD-PAD#3]
Hello! How can I make inference using a provided model/models? How to load pretrained models? Could you provide an example?

**Comment by dimasheva1** [comment]
I have tried to make an inference:

```
from model.mfad import FAD_HAM_Net
import torch

net = FAD_HAM_Net(pretrain=True, variant='resnet50')
net.load_state_dict(torch.load("checkpoints/LMFD_PAD_models/icm_o.pth", map_location=torch.device('cpu')))
net.eval()

dummy_input = torch.ones((1, 3, 224, 224))
pred, _ = net(dummy_input)
print(pred)
```
But I have got a tensor as a result with a number -6.3298. How I can to convert it to the label?

**Comment by dimasheva1** [comment]
Can I use the sigmoid to get a resulting label?

**Comment by meilfang** [comment]
Hello, using sigmoid can only get score between 0 and 1. However, when we classify an image, a threshold was computed from development set and then used for testing. If you only want to use those pretrained weight to test several images, I can provide the threshold of each model we used. If you want do evaluate on some datasets, you need to split dataset into development and test sets.

**Comment by dimasheva1** [comment]
> Hello, using sigmoid can only get score between 0 and 1. However, when we classify an image, a threshold was computed from development set and then used for testing. If you only want to use those pretrained weight to test several images, I can provide the threshold of each model we used. If you want do evaluate on some datasets, you need to split dataset into development and test sets.

Thank you for the answer. Actually, I want to use those pretrained weight to test several images. As I correctly understand, I have to compare the number -6.3298 with the threshold. I'll be very grateful if you provide the threshold of each model you used.

**Comment by dimasheva1** [comment]
Could you answer the following questions?

1. Should I put a cropped face as an input? Or Do I have to put a full image as an input?
2. Is it required to do transformations that you used for the validation? F.e.
   ```
   from dataset import TestDataAugmentation

   image_x = cv2.imread('real.png')
   image_x = cv2.cvtColor(image_x, cv2.COLOR_BGR2RGB)
   image_x = composed_transforms(image=image_x)['image']
   image_x = image_x[np.newaxis, :]
   pred, _ = net(image_x)
   ```
   Or I can use something like that?
   ```
   image = cv2.imread('real.png')
   images = [image]
   blob = cv2.dnn.blobFromImages(images, 1.0 / 127.5, (224, 224), (127.5, 127.5, 127.5), swapRB=True)
   blob = torch.Tensor(blob)
   pred, _ = net(blob)
   ```
3. As I correctly understand, score>threshold mean that a face on a picture is real. Am I right?

**Comment by meilfang** [comment]
Hi, 
1. First, you should add 'torch.sigmoid(pred)' to get a score between 0-1. And then compare with the thresholds. if pred>threshold, this image is real. The thresholds we used are 0.613508, 0.6971898 , 0.7309441  and 0.53312653  for omi_c, ocm_i, icm_o, oci_m models, respectively.
2. yes, the input face is cropped face
3. the augmentation used during test is for normalizing image by ImageNet mean and std and resizing image to 224x224x3 (please check TestDataAugmentation in dataset.py). You can use other ways, just make sure it has same effect:)

**Comment by dimasheva1** [comment]
Thank you! You helped me a lot. Good luck)

**Comment by fisakhan** [comment]
@meilfang , @dimasheva1 can you please provide a simple script that combines all these steps. input: image, output: spoof or not spoof ? I'll be very thankful to you.

**Comment by cw9981** [comment]
```python

net = FAD_HAM_Net(pretrain=True, variant='resnet50')
net.load_state_dict(torch.load("icm_o.pth", map_location=torch.device('cpu')))
net.eval()

from PIL import Image
from torchvision import transforms
img = Image.open(filename).convert('RGB')
tfms = transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])])
img_tensor = tfms(img).to('cpu').unsqueeze(0)
pred, _ = net(img_tensor)
score = torch.sigmoid(pred)
score = score.detach().numpy()
print('pred: ', pred)
print('score: ', score)
```

Output:

```python
pred:  tensor([[-0.5228]], grad_fn=<MmBackward>)
score:  [[0.3722082]]
```


@fisakhan  give it a try

**Comment by fisakhan** [comment]
@meilfang suggested blob = cv2.dnn.blobFromImages(images, 1.0 / 127.5, (224, 224), (127.5, 127.5, 127.5), swapRB=True) and you @cw9981 used transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])]). Why? How/where did you get these values?
Whats wrong in the code I wrote earlier? I would like confirmation.

**Comment by cw9981** [comment]
@fisakhan  Preprocess input data should be the same as training.  check dataset.py

**Comment by fisakhan** [comment]
Thanks @cw9981. The method always outputs "spoof" even for live faces. I tested on many "live" faces and it returns "spoof" all the time. Do you agree?

**Comment by cw9981** [comment]
@fisakhan  the model  "icm_o.pth"  + face detector -  scrfd from insightface  + webcam from my mac +  perfect frontal face . it works nicely on me, but not so well on my friends.

**Comment by fisakhan** [comment]
I'm testing it on a publicly available dataset and will share results with you. The following method looks promising but not completely opensource/free.
https://github.com/DoubangoTelecom/FaceLivenessDetection-SDK

**Comment by fisakhan** [comment]
How do you crop the detected face before resizing it to 224x224 ?

**Comment by fisakhan** [comment]
icm_o.pth on CASIA-SURF dataset generate the following numbers:

ACCURACY | TPR | FPR | TNR | FNR
0.527            | 0.586 | 0.607 | 0.392 | 0.414

True Positive (TP) = when the image is spoof and this method classifies that as spoof
False Positive (FP) = when the image is live and this method classifies that as spoof

**Comment by meilfang** [comment]
@cw9981  Thank you for your help in replying!
@fisakhan Sorry for the later reply and thank you for your interest in this work. 
1) I didnt suggest using blob = cv2.dnn.blobFromImages(images, 1.0 / 127.5, (224, 224), (127.5, 127.5, 127.5), swapRB=True). just dimasheva1 asked if ok to use it. During test the image should be normalized  by ImageNet mean and std and resizing image to 224x224x3 which also used in training as I and cw9981 answered before.
2) when you use those pretrained models for testing and after getting prediction score, do you use the threshold to decide if an image is attack or not?
4) I use MTCNN to detect face (please check paper), and then resize to 256x256 and then use center crop (please check test.py)
3) for results on CASIA-SURF: sorry for that the results looks not promising. I also have this database. I will also try to test on it and will let you know might after new year.
If you have further questions, please let me know, I will answer as soon as possible.
